File: graphtransliterator/process.py
# -*- coding: utf-8 -*-
"""
Functions to process "easy reading" settings of GraphTransliterator.
"""

import re


# ----------- process settings (main method) ----------


def _process_easyreading_settings(settings):
    """
    Extract information from "easy reading" settings.

    Returns
    -------
    `dict` keyed by:
      `tokens`: dict of {str: list of str}
      `rules`:  list of dict {'prev_classes', 'prev_tokens', 'tokens',
                'next_tokens', 'next_classes'}
      `whitespace': dict {'default', 'token_class', 'consolidate'}
      `onmatch_rules`: list of dict {
                       'prev_classes', 'next_classes', 'production'}
    """

    return {
        "tokens": settings.get("tokens"),
        "rules": _process_rules(settings.get("rules")),
        "onmatch_rules": _process_onmatch_rules(settings.get("onmatch_rules")),
        "whitespace": settings.get("whitespace"),
        "metadata": settings.get("metadata", {}),
    }


# ----------- process rules ----------

# added \n and \r explicitly as validator does not accept re.DOTALL

RULE_RE = re.compile(
    r"(?:\(((?:\s?<.+?>\s+)+)?(.+?)\s?\) |((?:\s?<.+?>\s+)+)?)?"
    r"((?:\n|\r|.)+?)"
    r"(?:"
    r"\s+(?:\((.+?)((?:\s+<.+?>?)?)\)$)"
    r"|"
    r"((?:\s+<.+?>)+)"
    r"|"
    r"$)"
)


def _process_rule(key, value):
    """Convert key and value of a "easy reading" rule into a dict."""

    rule = {"production": value}  # needs to be dict for Cerberus

    match = RULE_RE.match(key)
    assert match, "Rule pattern is not valid: %s" % key

    if match.group(1):
        rule["prev_classes"] = re.findall("<(.+?)>", match.group(1))
    if match.group(2):
        rule["prev_tokens"] = match.group(2).split(" ")
    if match.group(3):
        rule["prev_classes"] = re.findall("<(.+?)>", match.group(3))
    if match.group(4) == " ":
        rule["tokens"] = [" "]
    else:
        rule["tokens"] = [x for x in match.group(4).split(" ") if x != ""]
    if match.group(5):
        rule["next_tokens"] = match.group(5).split(" ")
    if match.group(6):
        rule["next_classes"] = re.findall("<(.+?)>", match.group(6))
    if match.group(7):
        rule["next_classes"] = re.findall("<(.+?)>", match.group(7))

    return rule


def _process_rules(easyreading_rules):
    """Processes "easy reading" rules dict into a list of dict."""

    return [_process_rule(key, value) for key, value in easyreading_rules.items()]


# ----------- process onmatch_rules -----------


ONMATCH_RE = re.compile(
    r"^(" r"(?:<[^+< \s]+>\s*)+" r")" r"\+" r"(" r"(?:\s*<[^+<]+>)+)\s*$"
)

ONMATCH_CLASS_RE = re.compile(r"(?<=<)[^+< \s]+(?=>)")


def _process_onmatch_rules(easyreading_onmatch_rules):
    """Process "easyreading" onmatch_rules (list of dict) into list of dict."""

    if not easyreading_onmatch_rules:
        return []

    processed_onmatch_rules = []

    for item in easyreading_onmatch_rules:
        assert len(item) == 1, "onmatch_rule has too many values: %s" % item

        (rule_easyreading, _production) = list(item.items())[0]

        match = ONMATCH_RE.match(rule_easyreading)

        prev_str = match.group(1)
        next_str = match.group(2)

        _prev_classes = ONMATCH_CLASS_RE.findall(prev_str)
        _next_classes = ONMATCH_CLASS_RE.findall(next_str)

        onmatch_rule_processed = {
            "prev_classes": _prev_classes,
            "next_classes": _next_classes,
            "production": _production,
        }

        processed_onmatch_rules.append(onmatch_rule_processed)

    return processed_onmatch_rules

File: packages/graphtransliterator-ts/src/process.ts
// src/process.ts

import {
    RawRuleDict,
    RawOnMatchDict,
    RawWhitespaceDict,
    EasyReadingInput,
    ProcessedSettingsDict,
} from "./types.js";

// ----------- Regex Patterns ----------

// Regex matching rule syntax key string
const RULE_RE =
    /^(?:\(((?:\s?<.+?>\s+)+)?(.+?)\s?\)\s|((?:\s?<.+?>\s+)+)?)?((?:\n|\r|.)+?)(?:\s+\((.+?)((?:\s+<.+?>?)?)\)$|((?:\s+<.+?>)+)|$)/;

// Regex matching onmatch rule syntax key string
const ONMATCH_RE = /^((?:<[^+< \s]+>\s*)+)\+((?:\s*<[^+<]+>)+)\s*$/;
const ONMATCH_CLASS_RE = /<([^+< \s]+)>/g;

// ----------- Helper Regex Extractors ----------

function matchAllAngleBrackets(str: string): string[] {
    const matches: string[] = [];
    const regex = /<(.+?)>/g;
    let match: RegExpExecArray | null;
    while ((match = regex.exec(str)) !== null) {
        matches.push(match[1]);
    }
    return matches;
}

// ----------- process rules ----------

/** Convert key and value of an "easy reading" rule into a dict. */
export function processRule(key: string, value: string): RawRuleDict {
    const rule: RawRuleDict = { production: value, tokens: [] };

    const match = RULE_RE.exec(key);
    if (!match) {
        throw new Error(`Rule pattern is not valid: ${key}`);
    }

    // Groups:
    // 1: prev_classes inside parentheses
    // 2: prev_tokens inside parentheses
    // 3: prev_classes outside parentheses
    // 4: tokens
    // 5: next_tokens inside parentheses
    // 6: next_classes inside parentheses
    // 7: next_classes outside parentheses

    if (match[1]) {
        rule.prev_classes = matchAllAngleBrackets(match[1]);
    }
    if (match[2]) {
        rule.prev_tokens = match[2].split(" ");
    }
    if (match[3]) {
        rule.prev_classes = matchAllAngleBrackets(match[3]);
    }

    if (match[4] === " ") {
        rule.tokens = [" "];
    } else {
        rule.tokens = match[4].split(" ").filter((x) => x !== "");
    }

    if (match[5]) {
        rule.next_tokens = match[5].split(" ");
    }
    if (match[6]) {
        rule.next_classes = matchAllAngleBrackets(match[6]);
    }
    if (match[7]) {
        rule.next_classes = matchAllAngleBrackets(match[7]);
    }

    return rule;
}

/** Processes "easy reading" rules dict into a list of rule objects. */
export function processRules(
    easyreadingRules: Record<string, string>,
): RawRuleDict[] {
    return Object.entries(easyreadingRules).map(([key, value]) =>
        processRule(key, value),
    );
}

// ----------- process onmatch_rules -----------

/** Process "easyreading" onmatch_rules into list of dict. */
export function processOnMatchRules(
    easyreadingOnmatchRules?: Record<string, string>[] | null,
): RawOnMatchDict[] {
    if (!easyreadingOnmatchRules || easyreadingOnmatchRules.length === 0) {
        return [];
    }

    const processedOnmatchRules: RawOnMatchDict[] = [];

    for (const item of easyreadingOnmatchRules) {
        const entries = Object.entries(item);
        if (entries.length !== 1) {
            throw new Error(
                `onmatch_rule has too many values: ${JSON.stringify(item)}`,
            );
        }

        const [ruleEasyreading, production] = entries[0];
        const match = ONMATCH_RE.exec(ruleEasyreading);
        if (!match) {
            throw new Error(`Onmatch pattern is not valid: ${ruleEasyreading}`);
        }

        const prevStr = match[1];
        const nextStr = match[2];

        const prevClasses: string[] = [];
        let prevMatch: RegExpExecArray | null;
        ONMATCH_CLASS_RE.lastIndex = 0;
        while ((prevMatch = ONMATCH_CLASS_RE.exec(prevStr)) !== null) {
            prevClasses.push(prevMatch[1]);
        }

        const nextClasses: string[] = [];
        let nextMatch: RegExpExecArray | null;
        ONMATCH_CLASS_RE.lastIndex = 0;
        while ((nextMatch = ONMATCH_CLASS_RE.exec(nextStr)) !== null) {
            nextClasses.push(nextMatch[1]);
        }

        processedOnmatchRules.push({
            prev_classes: prevClasses,
            next_classes: nextClasses,
            production,
        });
    }

    return processedOnmatchRules;
}

// ----------- process settings (main method) ----------

/** Extract information from "easy reading" settings. */
export function processEasyReadingSettings(
    settings: EasyReadingInput,
): ProcessedSettingsDict {
    const rulesInput = (settings.rules as Record<string, string>) || {};
    const onmatchInput =
        (settings.onmatch_rules as Record<string, string>[]) || [];

    const tokensBlock = (settings.tokens as Record<string, string[]>) || {};
    const whitespaceBlock = (settings.whitespace as RawWhitespaceDict) || {};
    const metadataBlock = (settings.metadata as Record<string, any>) || {};

    return {
        tokens: tokensBlock,
        rules: processRules(rulesInput),
        onmatch_rules: processOnMatchRules(onmatchInput),
        whitespace: whitespaceBlock,
        metadata: metadataBlock,
    };
}
